### PR TITLE
[Feat] Fetch a complete postal code from a redacted UK/Canada postal code

### PR DESCRIPTION
**Pull request by akshay2000saxena**
### Problem
When the Apple/Google Pay sheet is loading, the payment sheet has access to a redacted postal code (the first three digits).

### Context
For Canada -> It has access to only the forward sortation area (FSA) of the postal code (the first three letters).
For UK -> It has access to only the outward code of the postal code (the first half of the postal code).

To get Shipping options, the full postal code is required.
We essentially need to generate a random valid postal code based off of the first three digits so that we can get shipping options, and then when the payment is authorized, we'll have access to the full postal code and then we update the buyer identity with the real postal code (so shipping info is always correct)

### Changes
In this PR, we use the following:

For Canada -> a pre-computed JSON from a CSV containing all zip codes: https://raw.githubusercontent.com/ccnixon/postalcodes/master/CanadianPostalCodes.csv

For UK -> an API called https://api.postcodes.io/postcodes

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -9,3 +14,110 @@ export function isGooglePayAddress(
...
+ * @param {string} url - The `url` parameter is a string that represents the URL from which the zip
+ * code data will be fetched. This URL should point to a CSV file that contains the zip code data.
+ * @param {string} prefix - The `prefix` parameter is a string that represents the desired prefix of
+ * the zip code. The function will search for the first zip code in the data that starts with this
+ * prefix.
+ * @returns The function `findFirstZipCode` returns a `Promise` that resolves to a `string` if a
+ * matching zip code is found, or `null` if no matching zip code is found.
+ */
+async function findFirstZipCode(url: string, prefix: string): Promise<string | null> {
+  return new Promise((resolve, reject) => {
```
Comment (akshay2000saxena) [inline src/utils.ts:61]: `papaparse` is designed as a callback-based library, hence the inconsistency regarding async/await

File: src/utils.ts [hunk 2]
```diff
@@ -9,3 +14,110 @@ export function isGooglePayAddress(
...
   return 'name' in address;
 }
+
+/***
+This function is required for Canada and UK for Apple Pay and Google Pay.
+
+Reason:
+When the Apple/Google Pay sheet is loading, the payment sheet has access to a redacted postal code (the first three digits).
+For Canada -> It has access to only the forward sortation area (FSA) of the postal code (the first three letters).
+For UK -> It has access to only the outward code of the postal code (the first half of the postal code).
```
Comment (mattdenkers-rye) [inline src/utils.ts:24]: What about for the US?

Comment (akshay2000saxena) [reply src/utils.ts:24]: US postal codes are not redacted

File: src/utils.ts [hunk 3]
```diff
@@ -9,3 +14,110 @@ export function isGooglePayAddress(
...
+ * the zip code. The function will search for the first zip code in the data that starts with this
+ * prefix.
+ * @returns The function `findFirstZipCode` returns a `Promise` that resolves to a `string` if a
+ * matching zip code is found, or `null` if no matching zip code is found.
+ */
+async function findFirstZipCode(url: string, prefix: string): Promise<string | null> {
+  return new Promise((resolve, reject) => {
+    Papa.parse(url, {
+      download: true,
+      complete: (results) => {
```
Comment (mattdenkers-rye) [inline src/utils.ts:64]: Does this mean we download the whole CSV, so all of the data is in memory, and then we search through?

File: src/utils.ts [hunk 4]
```diff
@@ -9,3 +14,110 @@ export function isGooglePayAddress(
...
+        const data = results.data as string[][];
+        for (const row of data) {
+          const zipCode = row[0];
+          if (zipCode && zipCode.startsWith(prefix)) {
+            resolve(zipCode);
+            return;
+          }
+        }
+        resolve(null); // No matching zip code found
```
Comment (mattdenkers-rye) [inline src/utils.ts:73]: Loving the documentation

File: src/utils.ts (hunk 2 above)
Comment (mattdenkers-rye) [reply src/utils.ts:24]: So this sentence

"When the Apple/Google Pay sheet is loading, the payment sheet has access to a redacted postal code (the first three digits)"

Is only true for Canada / UK? But for US we have the full zip?

File: src/utils.ts [hunk 5]
```diff
@@ -9,3 +14,110 @@ export function isGooglePayAddress(
...
+
+/**
+ * The function `getFullPostalCodeForCanada` retrieves the full postal code for a given postal code in
+ * Canada, using a CSV file of Canadian postal codes.
+ * @param {string} postalCode - The `postalCode` parameter is a string representing a partial postal code for Canada.
+ * @returns a Promise that resolves to a complete postal code string.
+ */
+async function getFullPostalCodeForCanada(postalCode: string): Promise<string> {
+  const canadianPostalCodesCsv =
+    'https://raw.githubusercontent.com/ccnixon/postalcodes/master/CanadianPostalCodes.csv';
```
Comment (mattdenkers-rye) [inline src/utils.ts:111]: Hmmm a couple ideas here:
1. Should we host this ourselves? Wouldn't want `ccnixon` to delete the file for some reason
2. Do we need all the data? If we only care about the zip codes themselves then we could reduce the file size a lot
3. If we're already planning to manipulate the file and host it ourselves, we could do JSON instead, then we don't even need papaparse

Thoughts?

File: src/utils.ts [hunk 6]
```diff
@@ -9,3 +14,110 @@ export function isGooglePayAddress(
...
+  const ldu =
+    getRandomChar(POSTAL_CODE_LETTERS) +
+    getRandomChar(POSTAL_CODE_DIGITS) +
+    getRandomChar(POSTAL_CODE_LETTERS); // Local Delivery Unit
+  const fallbackPostalCode = postalCode + ' ' + ldu;
```
Comment (mattdenkers-rye) [inline src/utils.ts:116]: same nit: we might as well only generate this when we need it

File: src/utils.ts (hunk 2 above)
Comment (akshay2000saxena) [reply src/utils.ts:24]: Yes, lemme update the comment to mention only Canada and UK

File: src/utils.ts (hunk 5 above)
Comment (foreverest) [reply src/utils.ts:111]: +1 to all of these

File: src/index.ts [hunk 7]
```diff
@@ -5,3 +5,7 @@ export * from './applePay';
 export * from './constants';
 export * from './types';
 export * from './utils';
+
+// Import the JSON file
+import canadaPostalCodes from './canadaPostalCodes.json';
+export { canadaPostalCodes };
```
Comment (mattdenkers-rye) [inline src/index.ts:11]: Why do this here?

**Review by mattdenkers-rye — APPROVED** [review]

**Pull request merged**